src/AvoStorage.ts

```diff
@@ -5,53 +5,129 @@ export class AvoStorage {
   AsyncStorage: any | null = null;
 
   shouldLog: boolean;
-  initialized = false;
-  onInitFuncs: Array<() => void> = [];
+  storageInitialized = false;
+  itemsFromLastSessionLoaded = false;
+  useFallback = false;
+  fallbackStorage: any = {};
+  onStorageInitFuncs: Array<() => void> = [];
+  onItemsFromLastSessionLoadedFuncs: Array<() => void> = [];
 
   constructor(shouldLog: boolean) {
     this.shouldLog = shouldLog;
     if (!process.env.BROWSER) {
       this.reactNative = require("react-native");
       this.Platform = this.reactNative.Platform;
       this.AsyncStorage = this.reactNative.AsyncStorage;
-      
 
       if (this.Platform.OS === "android") {
+        // Android is the only platform that requires async loading of items from the prev session
+        this.initializeAndroid();
         this.AsyncStorage.getAllKeys().then((keys: Array<any>) =>
           this.AsyncStorage.multiGet(keys).then(
             (keyVals: Array<Array<string>>) => {
               keyVals.forEach((keyVal) => {
                 let key = keyVal[0];
                 this.data[key] = keyVal[1];
               });
-              this.initialized = true;
-              this.onInitFuncs.forEach((func) => {
-                func();
-              });
+              this.itemsLoadedAndroid();
```

Comment:
Here loading items from prev session might contain items from current session, if we initialize before `itemsLoadedAndroid` is done, is that ok?

Reply:
any reason to not just go with one `initializeAndItemsLoadedAndroid` in the end?

Reply:
> Here loading items from prev session might contain items from current session, if we initialize before itemsLoadedAndroid is done, is that ok?
Should be fine, we need all the data
> any reason to not just go with one initializeAndItemsLoadedAndroid in the end?
The thing that worried me in the previous implementation is that on android getItemAsync would return null until `this.AsyncStorage.getAllKeys().then` is fired and it's not the desired behaviour.

Reply:
I'm looking at it now and it seems like android is actually using something very similar to the fallback storage

Reply:
This is the most complex part of the project right now :)

Reply:
There is only one place using `getItemAsync`, we can probably get rid of it and simplify the storage implementation

Reply:
Yeah I think we should use only one flag, `storageInitialized`, logically it makes sense.